rotkehlchen/chain/solana/modules/jupiter/decoder.py

```diff
@@ -30,118 +27,153 @@
+        # Combine any opposite side events that have the same asset
+        events_to_skip = set()
+        for in_event in in_events_by_asset.values():
+            if (out_event := out_events_by_asset.get(in_event.asset)) is None:
+                continue
 
-            filtered_events.append(event)
+            if in_event.amount == out_event.amount:
+                events_to_skip.add(in_event)
+                events_to_skip.add(out_event)
+            elif out_event.amount > in_event.amount:
+                out_event.amount -= in_event.amount
+                events_to_skip.add(in_event)
+            else:  # out_event.amount < in_event.amount
+                in_event.amount -= out_event.amount
+                events_to_skip.add(out_event)
 
-        # replace decoded events with only the filtered events, removing any intermediate swaps
-        # so there is only a single swap for the full route.
-        context.decoded_events[:] = filtered_events
+        # Update the amount in the event notes and skip unneeded events
+        out_events: list[SolanaEvent] = []
+        in_events: list[SolanaEvent] = []
+        slippage_events: list[SolanaEvent] = []
+        for events_by_asset, final_list, sub_type, notes_template in (
+            (out_events_by_asset, out_events, HistoryEventSubType.SPEND, 'Swap {amount} {asset} in Jupiter'),  # noqa: E501
+            (in_events_by_asset, in_events, HistoryEventSubType.RECEIVE, 'Receive {amount} {asset} as the result of a swap in Jupiter'),  # noqa: E501
+        ):
+            for trade_event in events_by_asset.values():
+                if trade_event in events_to_skip:
+                    continue
 
-        if out_event is None or in_event is None:
+                trade_event.counterparty = CPT_JUPITER
+                if (
+                    sub_type == HistoryEventSubType.RECEIVE and
+                    destination_mint not in trade_event.asset.identifier
+                ):
+                    trade_event.event_type = HistoryEventType.RECEIVE
+                    trade_event.event_subtype = HistoryEventSubType.NONE
+                    trade_event.notes = f'Receive {trade_event.amount} {trade_event.asset.resolve_to_asset_with_symbol().symbol} due to positive slippage in a Jupiter swap'  # noqa: E501
+                    slippage_events.append(trade_event)
+                    continue
+
+                trade_event.event_type = HistoryEventType.TRADE
+                trade_event.event_subtype = sub_type
+                trade_event.notes = notes_template.format(
+                    amount=trade_event.amount,
+                    asset=trade_event.asset.resolve_to_asset_with_symbol().symbol,
+                )
+                final_list.append(trade_event)
+
+        # replace decoded events with only the filtered events
+        filtered_events = (
+            out_events + in_events +
+            ([platform_fee_event] if platform_fee_event is not None else []) +
```

Comment:
if you pass this directly to `maybe_reshuffle_events`, you won't need this inline if/else